### PR TITLE
Newfilestructure

### PR DESCRIPTION
**Pull request by Msalter91**
All post controls work in an object oriented way and all follow PSR-12. The only change I still need to make for compliance is to create a namespace for all classes but this will take some time.

### CODE REVIEW

**Comment by szreka** [comment]
Hello

File: app/code/Post/Controllers/Index.php [hunk 1]
```diff
@@ -0,0 +1,157 @@
+<?php
+
+declare(strict_types=1);
+
+use library\Controller as Controller;
+
+class Index extends Controller
+{
+    protected ?object $postModel = null;
+
```
Comment (cfuzek) [inline app/code/Post/Controllers/Index.php:10]: Please Remove the space

File: app/code/User/Api/Data/UserInterface.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+    public function getEmail() :string;
+    public function setEmail(string $email);
+
+    public function getPassword() :string;
+    public function setPassword(string $password);
+
+    public function getCreatedAt() :string;
+    public function setCreatedAt(string $datetime);
+
```
Comment (cfuzek) [inline app/code/User/Api/Data/UserInterface.php]: Please remove the space

File: app/code/User/Api/UserRepositoryInterface.php [hunk 3]
```diff
@@ -0,0 +1,13 @@
+<?php
+
```
Comment (cfuzek) [inline app/code/User/Api/UserRepositoryInterface.php:2]: Please remove the space

File: app/code/User/Api/UserRepositoryInterface.php [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+declare(strict_types=1);
+
+interface UserRepositoryInterface
+{
+    public function getById(int $id): User;
+
+    public function save(User $user): ?User;
+
```
Comment (cfuzek) [inline app/code/User/Api/UserRepositoryInterface.php]: Please remove the space

File: app/code/User/Controllers/Users.php [hunk 5]
```diff
@@ -1,34 +1,34 @@
...
     protected ?object $userModel = null;
 
     public function __construct()
     {
         $this->userModel = $this->model('User');
     }
-    public function createUserSession($user){
-        //TODO Writing strings (poss 255 chars) to three session variables will be slow! Can this be replaced with a Bool?
+
+    public function createUserSession($user)
```
Comment (cfuzek) [inline app/code/User/Controllers/Users.php]: Please add the type declaration for the argument type

File: app/code/User/Controllers/Users.php [hunk 6]
```diff
@@ -40,90 +40,68 @@ public function register(){
...
 
-                // Register the user using the Model
-
-                if($this->userModel->register($data)){
+                if ($this->userModel->register($data)) {
                     flash('register_success', 'You are now registered');
-                    redirect('users/login');
+                    redirect('user/users/login');
                 } else {
                     die('Something has gone wrong');
```
Comment (cfuzek) [inline app/code/User/Controllers/Users.php]: We should not use die in the code. Please review a different solution

File: app/code/User/Controllers/Users.php [hunk 7]
```diff
@@ -40,90 +40,68 @@ public function register(){
...
+                $data['password_error'] = 'Password much be at least 6 characters';
             }
-
-            if($this->userModel->findUserByEmail($data['email'])){
+            if ($this->userModel->findUserByEmail($data['email'])) {
                 $loggedInUser = $this->userModel->login($data['email'], $data['password']);
 
-                if($loggedInUser){
+                if ($loggedInUser) {
                     //Create session
```
Comment (cfuzek) [inline app/code/User/Controllers/Users.php]: Please remove the comment

File: app/code/User/Controllers/Users.php [hunk 8]
```diff
@@ -136,13 +114,13 @@ public function login(){
                 $data['email_error'] = 'No user Found';
             }
 
-            if(!empty($data['email_error']) || !empty($data['password_error'])) {
+            if (!empty($data['email_error']) || !empty($data['password_error'])) {
                 $this->view('users/login', $data);
             }
         } else {
             // Load the form
```
Comment (cfuzek) [inline app/code/User/Controllers/Users.php]: Please remove the comment

File: app/code/User/Model/User.php [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+{
+    private ?Database $db = null;
+    private array $data = [];
+
+    public function __construct()
+    {
+        $this->db = new Database();
+    }
+
+    public function register($data): bool
```
Comment (cfuzek) [inline app/code/User/Model/User.php]: You should use the argument type and $data should be an object of the class User

File: app/code/User/Model/User.php [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+        $this->db->bind(':password', $data['password']);
+
+        if ($this->db->execute()) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    public function login($email, $password): bool | object
```
Comment (cfuzek) [inline app/code/User/Model/User.php]: Argument type missing

File: app/code/User/Model/User.php [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+        $row = $this->db->singleResult();
+
+        if ($this->db->rowCount() > 0) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    public function getUserById($id): bool | object
```
Comment (cfuzek) [inline app/code/User/Model/User.php]: Argument type missing

File: app/code/User/Model/User.php [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    public function getName(): string
+    {
+        // TODO: Implement getName() method.
+    }
+
+    public function setUserId(string $name)
+    {
+        // TODO: Implement setUserId() method.
```
Comment (cfuzek) [inline app/code/User/Model/User.php:85]: Please complete the implementation

File: app/code/User/Model/User.php [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+
+        if (password_verify($password, $hashed_password)) {
+            return $row;
+        } else {
+            return false;
+        }
+    }
+
+
+    public function findUserByEmail($email): bool
```
Comment (cfuzek) [inline app/code/User/Model/User.php]: Argument type missing

File: app/code/Post/Controllers/Index.php [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+     * @returns null
+     */
+    public function delete($id): void
+    {
+        if ($_SERVER['REQUEST_METHOD'] == 'POST') {
+            $getRepositoryPost = new PostRepository();
+            $post = $getRepositoryPost->getById(intval($id));
+
+            $postid = $post->getUserId();
+            var_dump($_SESSION['user_id']);
```
Comment (cfuzek) [inline app/code/Post/Controllers/Index.php]: Please remove the debug

**Review by cfuzek — CHANGES_REQUESTED** [review]

File: app/code/User/Controllers/Users.php [hunk 15]
```diff
@@ -1,34 +1,34 @@
...
 
-// TODO Remove the dynamically declared properties
-// TODO Property and return types
+use library\Controller as Controller;
 
-{
+class Users extends Controller
 
+{
     protected ?object $userModel = null;
```
Comment (cfuzek) [inline app/code/User/Controllers/Users.php:11]: test

File: app/code/User/Model/User.php (hunk 9 above)
Comment (Msalter91) [reply app/code/User/Model/User.php]: We were only working on the posts part of the model so the user still mostly follows the old structure. I can work on the user throughout the week to bring it in line.

File: app/code/Post/Controllers/Index.php [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+        $this->view('posts/index', $data);
+    }
+
+    /**
+     * @return void
+     */
+    public function add(): void
+    {
+        $post = new Post();
+        if ($_SERVER['REQUEST_METHOD'] == 'POST') {
```
Comment (rszekely) [inline app/code/Post/Controllers/Index.php:33]: Try an early return here, so ie:
```
 if ($_SERVER['REQUEST_METHOD'] != 'POST') {
   $this->view('posts/add', $post);
   return;
} 
```

Comment (rszekely) [reply app/code/Post/Controllers/Index.php:33]: If you like it, you could try refactoring the whole class.

**Pull request merged**